src/transport/types.ts

```diff
@@ -0,0 +1,120 @@
+import events from 'events'
+
+/**
+ * A libp2p transport is understood as something that offers a dial and listen interface to establish connections.
+ */
+export interface Interface {
+  /**
+   * Dial a given multiaddr.
+   *
+   * @param {Multiaddr} ma
+   * @param {any} [options]
+   * @returns {Promise<Connection>}
+   */
+  dial(ma: Multiaddr, options?: any): Promise<Connection>;
+  /**
+   * Create transport listeners.
+   *
+   * @param {any} options
+   * @param {(Connection) => void} handler
+   */
+  createListener(options: any, handler: (Connection) => void): Listener;
+  /**
+   * Takes a list of `Multiaddr`s and returns only valid addresses for the transport
+   *
+   * @param {Multiaddr[]} multiaddrs
+   * @returns {Multiaddr[]}
+   */
+  filter(multiaddrs: Multiaddr[]): Multiaddr[];
+}
+
+export interface Listener extends events.EventEmitter {
+  /**
+   * Start a listener
+   *
+   * @param {Multiaddr} multiaddr
+   * @returns {Promise<void>}
+   */
+  listen(multiaddr: Multiaddr): Promise<void>;
+  /**
+   * Get listen addresses
+   *
+   * @returns {Multiaddr[]}
+   */
+  getAddrs(): Multiaddr[];
+  /**
+   * Close listener
+   *
+   * @returns {Promise<void>}
+   */
+  close(): Promise<void>;
+}
+
+export interface Upgrader {
+  /**
+   * Upgrades an outbound connection on `transport.dial`.
+   *
+   * @param {MultiaddrConnection} maConn
+   * @returns {Promise<Connection>}
+   */
+  upgradeOutbound(maConn: MultiaddrConnection): Promise<Connection>;
+
+  /**
+   * Upgrades an inbound connection on transport listener.
+   *
+   * @param {MultiaddrConnection} maConn
+   * @returns {Promise<Connection>}
+   */
+  upgradeInbound(maConn: MultiaddrConnection): Promise<Connection>;
+}
+
+export declare class Transport implements Interface {
```

Comment:
I think this should be an interface as well not a class.

Reply:
> if the purpose here is to type the constructor

It is and it would work for the transports. However js-libp2p uses creates its configured transports as `new Transport()` and not `new TCP()`, so the type there should be `Transport`. I think we can go with the base class in a follow up PR, as this will need all the transports to also be changed. 
We are targeting a first iteration with just support in the core.

I will use the interface for now

Reply:
I think it is important to distinguish between type (interface) of the value and how do you create one. In most places you'd require value of certain type to be provided regardless how it was created & in much fewer places (if any) you would require a factory for creating these values. The fact that factory happens to be the class / constructor for the type should really be irrelevant.

That is to suggest that we should separate those concerns and decouple them from each other even if in practice both are implemented via same class definition.

Reply:
@hugomrdias found a way of adding the constructor in the interface. So this seems good now just with the interface